File: DesignAndAnalysisOfAlgorithm/DivideAndConquerAndGreedMethods/QUICKSORT/QuickSort.py
arr=[50,1,77,6,2,9,7,1,0]
def QuickSort(arr):
   if(len(arr)<=1):
      return arr
   pivot=arr[len(arr)-1]
   p=0
   for j in range(len(arr)-1):
      if(arr[j]<=pivot):
         arr[p],arr[j]=arr[j],arr[p]
         p+=1
   arr[p],arr[len(arr)-1]=arr[len(arr)-1],arr[p]
   less=arr[:p]
   greater=arr[p+1:]
   return QuickSort(less)+[pivot]+QuickSort(greater)
print(QuickSort(arr))


def QuickSort(arr):
   if len(arr)<=1:
      return arr
   pivot=arr[0]
   less=[x for x in arr[1:] if x<=pivot]
   greater=[x for x in arr[1:] if x>pivot]
   return QuickSort(less)+[pivot]+QuickSort(greater)
print(QuickSort(arr))

File: DesignAndAnalysisOfAlgorithm/DivideAndConquerAndGreedyMethods/Divide&Conquer/Sorting/QuickSort/QUICKSORT.java
public class QUICKSORT {
   static int Partition(int []arr){
      int pivot=arr[arr.length-1];
      int p=0;
      for(int i=0;i<arr.length-1;i++){
         int temp=arr[p];
         arr[p]=arr[i];
         arr[i]=temp;
         p++;
      }
      int temp=pivot;
      pivot=arr[p];
      arr[p]=temp;
      return p;
   }
   static int [] QuickSort(int []arr){
      if(arr.length<=1){
         return arr;
      }
      int p=Partition(arr);
      int[] less = new int[p];
      int[] greater = new int[arr.length - p-1];
      for (int i = 0; i < p; i++) {
        less[i] = arr[i];
      }
      QuickSort(less);
      for (int j = p+1; j < arr.length; j++) {
         greater[j - p+1] = arr[j];
      }
      QuickSort(greater);
      int [] array =new int[less.length+greater.length];
      for(int i=0;i<array.length;i++){
         int c=0;
         for (int j = 0; j < p; j++) {
            array[c] = arr[j];
            c++;
         }
         for (int j = 0; j < p; j++) {
            array[c] = arr[j];
            c++;
         }
      }
   return array;
   }
   public static void main(String[] args) {
      int[] arr = {50, 1, 7, 6, 2, 9, 77, 1, 0};
       QUICKSORT ob = new QUICKSORT();
       int[] sortedArray = ob.QuickSort(arr);
       for (int num : sortedArray) {
           System.out.print(num + " ");
       }
   }
 
}